org.spoofax.terms/src/org/spoofax/terms/StrategoArrayList.java

```diff
@@ -1,289 +1,96 @@
 package org.spoofax.terms;
 
+import org.spoofax.interpreter.terms.IStrategoArrayList;
 import org.spoofax.interpreter.terms.IStrategoList;
 import org.spoofax.interpreter.terms.IStrategoTerm;
 import org.spoofax.interpreter.terms.ITermPrinter;
 import org.spoofax.terms.attachments.ITermAttachment;
-import org.spoofax.terms.attachments.TermAttachmentType;
-import org.spoofax.terms.util.TermUtils;
-import java.io.IOException;
-import java.util.Arrays;
-import java.util.Collection;
-import java.util.Iterator;
-import java.util.List;
-import java.util.NoSuchElementException;
-import java.util.RandomAccess;
-
-import javax.annotation.Nullable;
 
-import static org.spoofax.terms.AbstractTermFactory.EMPTY_TERM_ARRAY;
+import java.io.IOException;
 
-public class StrategoArrayList extends StrategoTerm implements IStrategoList, RandomAccess {
+/**
+ * @see IStrategoArrayList for most of the shared implementation for this class and {@link StrategoArrayListTail}
+ */
+public class StrategoArrayList extends StrategoTerm implements IStrategoArrayList {
```

Comment:
If possible, could these classes be `package private`? Then they can only be instantiated and cast by the `TermBuilder`, but not any code outside the package. This is desirable, because it avoids that code can depend on the actual implementation of terms.